File: roveagent/workforce/personas_test.py
"""Executive personas map to the single RoveAgent runtime employees."""
from __future__ import annotations

import unittest

from roveagent.workforce.employees import find_employee
from roveagent.workforce.personas import PERSONA_ALIASES, persona_for


class PersonasTest(unittest.TestCase):
    def test_persona_aliases_map_to_employees(self) -> None:
        self.assertEqual(find_employee("ceo-insight").key, "ceo")
        self.assertEqual(find_employee("coo").key, "operations")
        self.assertEqual(find_employee("cmo").key, "marketing")
        self.assertEqual(find_employee("cto").key, "devops")
        # 旧 employee key 仍可直接使用
        self.assertEqual(find_employee("ceo").key, "ceo")
        self.assertIsNone(find_employee("nope"))

    def test_persona_metadata(self) -> None:
        cmo = persona_for("cmo")
        self.assertIsNotNone(cmo)
        self.assertEqual(cmo["employee_key"], "marketing")
        self.assertIn("customer growth", str(cmo["focus"]))
        cto = persona_for("cto")
        self.assertIn("system health", str(cto["focus"]))
        self.assertIn("cto", PERSONA_ALIASES)


if __name__ == "__main__":
    unittest.main()

File: tests/personas.test.ts
import test from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { PERSONAS, resolvePersonaKey } from '../src/lib/agent/personas';

/**
 * 记录一次正当变更（2026-09-25）：persona 由 4 个变为 5 个。
 *
 * 原断言写死 `PERSONAS.length === 4`，而运行时 `capability_router.py` 里
 * `developer`（file/terminal/git/skills 工具集）一直存在，前端却没有任何 persona
 * 映射到它 —— 用户实测「这个不能编码啊」时拿到的是 CEO Agent 的
 * "I don't have access to a terminal or file system"（CEO 本就不碰文件与终端）。
 *
 * 因此新增 Developer persona。下面不再写死数字，而是锁定**精确的 key 集合与映射**
 * ——这样"少一个"和"多一个"都会红，且不依赖魔法数字。
 */
test('executive personas exist with stable employee mapping', () => {
  const byKey = new Map(PERSONAS.map((p) => [p.key, p]));
  assert.deepEqual(
    [...byKey.keys()].sort(),
    ['ceo-insight', 'cmo', 'coo', 'cto', 'developer'],
    'persona 集合发生变化时必须显式更新本断言（缺失或多出都要可见）',
  );
  assert.equal(PERSONAS.length, byKey.size, 'persona key 不得重复');
  assert.equal(byKey.get('ceo-insight')?.employeeKey, 'ceo');
  assert.equal(byKey.get('coo')?.employeeKey, 'operations');
  assert.equal(byKey.get('cmo')?.employeeKey, 'marketing');
  assert.equal(byKey.get('cto')?.employeeKey, 'devops');
  assert.equal(byKey.get('developer')?.employeeKey, 'developer');
});

test('resolvePersonaKey normalizes unknown values to ceo-insight', () => {
  assert.equal(resolvePersonaKey('cmo'), 'cmo');
  assert.equal(resolvePersonaKey('cto'), 'cto');
  assert.equal(resolvePersonaKey(undefined), 'ceo-insight');
  assert.equal(resolvePersonaKey('hacker'), 'ceo-insight');
});

test('chat route forwards persona as employee key and never trusts raw input', () => {
  const source = readFileSync(join(process.cwd(), 'src/app/api/agent/chat/route.ts'), 'utf8');
  assert.match(source, /resolvePersonaKey/);
  assert.match(source, /PERSONA_EMPLOYEE/);
});

test('agent page exposes persona selector chips', () => {
  const source = readFileSync(join(process.cwd(), 'src/app/[locale]/agent/page.tsx'), 'utf8');
  assert.match(source, /PERSONAS\.map/);
  assert.match(source, /setPersona/);
  assert.match(source, /persona,/);
});

test('python personas map to the single runtime workforce', () => {
  const source = readFileSync(join(process.cwd(), 'roveagent/workforce/personas.py'), 'utf8');
  for (const key of ['ceo-insight', 'coo', 'cmo', 'cto', 'developer']) {
    assert.ok(source.includes('"' + key + '"'), key + ' persona missing');
  }
});
